datafusion/physical-plan/src/joins/stream_join_utils.rs

```diff
@@ -740,20 +617,338 @@ pub fn record_visited_indices<T: ArrowPrimitiveType>(
     }
 }
 
+#[macro_export]
+macro_rules! handle_state {
+    ($match_case:expr) => {
+        match $match_case {
+            Ok(StreamJoinStateResult::Continue) => continue,
+            Ok(StreamJoinStateResult::Ready(res)) => Poll::Ready(Ok(res).transpose()),
+            Err(e) => Poll::Ready(Some(Err(e))),
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! handle_async_state {
+    ($state_func:expr, $cx:expr) => {
+        $crate::handle_state!(ready!($state_func.poll_unpin($cx)))
+    };
+}
+
+/// Represents the possible results of a state in the join stream.
+pub enum StreamJoinStateResult<T> {
+    Ready(T),
+    Continue,
+}
+
+/// Represents the various states of an eager join stream operation.
+///
+/// This enum is used to track the current state of streaming during a join
+/// operation. It provides indicators as to which side of the join needs to be
+/// pulled next or if one (or both) sides have been exhausted. This allows
+/// for efficient management of resources and optimal performance during the
+/// join process.
+#[derive(Clone, Debug)]
+pub enum EagerJoinStreamState {
+    /// Indicates that the next step should pull from the right side of the join.
+    PullRight,
+
+    /// Indicates that the next step should pull from the left side of the join.
+    PullLeft,
+
+    /// State representing that the right side of the join has been fully processed.
+    RightExhausted,
+
+    /// State representing that the left side of the join has been fully processed.
+    LeftExhausted,
+
+    /// Represents a state where both sides of the join are exhausted.
+    ///
+    /// The `final_result` field indicates whether the join operation has
+    /// produced a final result or not.
+    BothExhausted { final_result: bool },
+}
+
+/// Represents the asynchronous trait for an eager join stream.
```

Comment:
Could we please define in these comments what an "eager join stream" is as I am not familiar with the concept and I think others may not be either? 

Perhaps we can include some types of joins that have this property  (I think Symmetric Hash Join and Merge Join?)

Is the key feature of an eager stream that it produces output as soon as possible as it reads input, rather than buffering? Is it true that an eager  join stream buffers as little data as possible, but potentially sacrifices some performance?

Reply:
Something else that might help would be to document some examples of how this is used (like what the control flow / state transitions could be). However, I don't have a great idea of how to do this specifically